Add symbolic tracing support to torch._dynamo.export (fake input + weights)

thiagocrepaldi opened the pull request:
Fixes #95900
Using the following repro as guide:

```python
import torch
import torch._dynamo
from torch._subclasses import fake_tensor
from torch.fx.experimental.symbolic_shapes import ShapeEnv
from torch._dynamo.output_graph import config
class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.linear = torch.nn.Linear(2, 2)
        self.linear2 = torch.nn.Linear(2, 2)

    def forward(self, x):
        out = self.linear(x)
        out = self.linear2(out)
        return out


fake_mode = fake_tensor.FakeTensorMode(allow_non_fake_inputs=False,
                                       allow_fallback_kernels=True,
                                       shape_env=ShapeEnv(
                                            allow_scalar_outputs=config.capture_scalar_outputs,
                                            allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
                                            frame_id=0
                                        ),
)
# Fakefying input/model before calling torch._dynamo.export
with fake_mode:
    fake_x = torch.rand(5, 2, 2)
    model = Model()

# Calling torch._dynamo.export without active fake mode
graph_module, guards = torch._dynamo.export(
    model,
    fake_x,
    aten_graph=True,
    fake_mode=fake_mode
)
graph_module.print_readable()
graph_module.graph.print_tabular()
```

Summary of changes:

    * Plumb fake_mode through torch.export API. When specified, it
    replaces the creation of a new FaketendorMode at InstructionTranslator on behalf of OutputGraph
     Hacks FakeTensor.__new__ to prevent a
    torch.tensor._make_subclass call for inputs that are already fakefied by
    user. This probably need to be fixed in a nicer way. Any idea?
    * Removed a few asserts that didn't want faked tensors coming
    from user script
    * Added torch._subclasses.fake_tensor.FakeTensor to type list on a few
    asserts check to allow fake inputs

The changes above allowed symbolic tracing with both static and dynamic shapes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @anijain2305 @soumith @desertfire

thiagocrepaldi commented:
https://github.com/pytorch/pytorch/issues/91720 might be the root cause for the static shapes

thiagocrepaldi commented:
@jansel @ezyang

Does `make_fx` support `tracing_mode="symbolic"`?

```python
import torch
import torch._dynamo
from torch.fx.experimental.symbolic_shapes import ShapeEnv
from torch._dynamo.output_graph import config
class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.linear = torch.nn.Linear(2, 2)
        self.linear2 = torch.nn.Linear(2, 2)

    def forward(self, x):
        out = self.linear(x)
        out = self.linear2(out)
        return out

model = Model()
x = torch.rand(2, 2)

from torch.fx.experimental import proxy_tensor
gm = proxy_tensor.make_fx(model, tracing_mode="symbolic", _allow_non_fake_inputs=True)(x)
gm.print_readable()
```

Currently it returns

```bash
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[2, 2]):
        # No stacktrace found for following nodes
        _param_constant0 = self._param_constant0
        t: f32[2, 2] = torch.ops.aten.t.default(_param_constant0);  _param_constant0 = None
        _param_constant1 = self._param_constant1
        addmm: f32[2, 2] = torch.ops.aten.addmm.default(_param_constant1, arg0_1, t);  _param_constant1 = arg0_1 = t = None
        _param_constant2 = self._param_constant2
        t_1: f32[2, 2] = torch.ops.aten.t.default(_param_constant2);  _param_constant2 = None
        _param_constant3 = self._param_constant3
        addmm_1: f32[2, 2] = torch.ops.aten.addmm.default(_param_constant3, addmm, t_1);  _param_constant3 = addmm = t_1 = None
        return addmm_1
```
which has static shape

Would this be related to https://github.com/pytorch/pytorch/issues/91720?

BowenBao commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -905,13 +905,16 @@ def device(self):
-            dispatch_device=True,
-            device_for_backend_keys=device,
-        )
+        if isinstance(elem, torch._subclasses.FakeTensor):
```
nit: don't recall, is this still needed?

thiagocrepaldi replied:
yup, otherwise it raises `RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor`

BowenBao replied:
In the end we might want to instead fix the callsite :)

BowenBao replied:
For context, we've found below as one of the callsite. The empty tensor was created as a fake tensor under user created fake mode context. It can be repro-ed shortly with
```python
with fake_mode:
    FakeTensor(fake_mode, torch.empty(...))
```

https://github.com/pytorch/pytorch/blob/14d0b6e0d8d2a087ff06ac7e5a1be5f4a64a7d0f/torch/_subclasses/fake_tensor.py#L779-L788

thiagocrepaldi replied:
would replace `self = torch.Tensor._make_subclass()` by `fake_mode.from_tensor()` an option?

thiagocrepaldi replied:
from fake tensor design doc:

> Fake tensor of fake tensor?
There is interest in sending fake tensors as user inputs into the PT2 stack, which would imply we would need to be able to create a fake tensor of a fake tensor. This isn’t really supported right now, but maybe it would not be too difficult to do.

I have no idea how to get around that in a different way. Ideas?

ezyang replied:
Didn't we agree that you were going to do the "make PT2 reuse the ambient fake tensor context" strategy

thiagocrepaldi replied:
@ezyang  We did! In fact, `torch/_dynamo/eval_frame.py::export` now takes a [fake_mode from user](https://github.com/pytorch/pytorch/pull/100017/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911R739) which is ultimately consumed in two places:

The first place is inside [OutputGraph.\_\_init\_\_](https://github.com/pytorch/pytorch/pull/100017/files#diff-1bbe80a42da83f0b0e1386c3aca92323930626a65173c7d7e838931446fb4324R229 and the second place is within [dynamo_normalization_capturing_compiler](https://github.com/pytorch/pytorch/pull/100017/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911R851)

Should this solve the issue mentioned here? Seems unrelated, because the user data will already be fakeifed and when the input is fakeifed again, internally, `FakeTensor.\_\_new\_\_` will be called and the issue `RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor` will still happen

thiagocrepaldi replied:
For reference, the issue with `make_subclass`comes from [these checks](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/python_variable.cpp#L1791-L1864) at `THPVariable_NewWithVar` which is called by `THPVariable_make_subclass`

eellison replied:
Do you have a small repro you could post of what the failing pattern is ?

ezyang replied:
So I think what is probably happening is something like, we are calling "detach()" to generate a fresh tensor object, but somehow the FakeTensor dispatch mode is on and interposing on the detach() call which fakeifies the object. We then are unable to create the new tensor without overwriting the old subclass.

thiagocrepaldi replied:
> Do you have a small repro you could post of what the failing pattern is ?

Yes, I do

```python
import torch

fake_mode = torch._subclasses.FakeTensorMode()
with fake_mode:
    fake_x = torch.randn(1,2,3)
```

which leads to

```bash
Traceback (most recent call last):
  File "/home/thiagofc/dev/pytorch_repros/fake_from_fake.py", line 7, in <module>
    fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)
  File "/home/thiagofc/dev/github/pytorch-dev1/torch/_subclasses/fake_tensor.py", line 904, in __new__
    self = torch.Tensor._make_subclass(
RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

eellison replied:
That doesn't error on master, it only errors with your changes. Do you have a small repro of what you are trying to that errors on master?

thiagocrepaldi replied:
Weird, I just tried with latest nightly and it repro'ed. It also reproed on CI before changing fake_tensor.py to create a fake tensor from another. here is some extra info

```bash
(dev2-cu113-py39) ✘-1 ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ conda list | grep -i torch
magma-cuda113             2.5.2                         1    pytorch
pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
torch                     2.1.0.dev20230518+cu118          pypi_0    pypi
torchaudio                2.1.0.dev20230518+cu118          pypi_0    pypi
torchvision               0.14.0a0+3c9ae0a           dev_0    <develop>
(dev2-cu113-py39) ✔ ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ python ~/dev/pytorch_repros/fake_from_fake.py 
Traceback (most recent call last):
  File "/home/thiagofc/dev/pytorch_repros/fake_from_fake.py", line 7, in <module>
    fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)
  File "/home/thiagofc/miniconda3/envs/dev2-cu113-py39/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 904, in __new__
    self = torch.Tensor._make_subclass(
RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
(dev2-cu113-py39) ✘-1 ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ cat ~/dev/pytorch_repros/fake_from_fake.py 
import torch

fake_mode = torch._subclasses.FakeTensorMode()
with fake_mode:
    fake_x = torch.randn(1,2,3)

fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)(dev2-cu113-py39) ✔ ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ 
```

eellison replied:
`fake_from_fake_x` wasn't in the repro. now i've got it!

eellison replied:
when would you want to use `fake_from_fake_x` instead of the existing `fake_x` ?

thiagocrepaldi replied:
We want to develop a brand new ONNX exporter (to replace the current TorchScript-based) which does symbolic tracing on model+input. This is important to export huge models

An example usage would be:

```python
# Fakefying input/model before calling torch._dynamo.export
with fake_mode:
    fake_x = torch.rand(5, 2, 2)
    fake_model = Model()

# Calling torch._dynamo.export without active fake mode
graph_module, guards = torch._dynamo.export(
    fake_model,
    fake_x,
    aten_graph=True,
    fake_mode=fake_mode
)
```

The full context is on the PR description, but let me know if anything is missing or confusing

ezyang replied:
@thiagocrepaldi but we shouldn't do this, right? If we're preserving the fake mode, we also want to preserve the input fake tensors, so we shouldn't end up fakeifying from an already fake tensor.

Actually this might be a bit troublesome, because it means if we mutate metadata on the fake tensors, that mutation will get reflected outside of the export. But we can figure that out elsewise.

eellison replied:
> Actually this might be a bit troublesome, because it means if we mutate metadata on the fake tensors, that mutation will get reflected outside of the export

IMO we should still be converting input FakeTensors to FakeTensors of a new mode. There are a couple reasons to do this :

1. Mutation as you have mentioned 
2. In dynamo tracing we trace with `allow_non_fake_inputs=False`. There have been many instances where we did fully unwrap a python call (see below) where our failure to convert inputs to fake tensors also revealed that we did not properly inline some piece of python logic and therefore add appropriate guards etc. I think we should have the same guardrails for fake tensor inputs. 


Example errors: https://github.com/pytorch/pytorch/issues/96411 and https://github.com/pytorch/pytorch/issues/96362

ezyang replied:
> IMO we should still be converting input FakeTensors to FakeTensors of a new mode.

I agree this is better but this is a biiiig refactor. Dynamo in particular makes hella assumptions everywhere.

Also, there are use cases for reusing the fake mode as is; for example, this is what @jansel wanted for the fx tracing for subgraph matching.

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -444,7 +444,7 @@ def getattr(self, attr, attr_val, parameter_proxy_cache):
 
     def create_arg(self, a: Any):
-        if isinstance(a, torch.nn.Parameter):
+        if isinstance(a, (torch.nn.Parameter, torch._subclasses.FakeTensor)):
```
The changes here look questionable

ezyang replied:
In particular, it doesn't make sense to have something that matches Parameter and FakeTensor, but not regular Tensor!

BowenBao replied:
Somehow weights of fakified model show up here as `torch._subclasses.FakeTensor` 🤔 . Normally they are `nn.Parameter`.

thiagocrepaldi replied:
> In particular, it doesn't make sense to have something that matches Parameter and FakeTensor, but not regular Tensor!

For this particular instance, it seems a faked `torch.nn.Parameter` is implemented as a `torch._subclasses.FakeTensor`. There is no `FakeParameter` or similar, which makes me think this change is ok. Makes sense?

thiagocrepaldi commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -240,7 +240,7 @@ def inner(e):
     return lambda t: get_proxy_slot(t, tracer, t)
 
-HANDLED_TYPES = (torch.Tensor, torch.nn.Parameter)
+HANDLED_TYPES = (torch.Tensor, torch.nn.Parameter, torch._subclasses.fake_tensor.FakeTensor)
```
@jansel @ezyang This line causes an interesting regression. It adds a bunch of unsued `detached` tensors that are not used

As an example, `test_simple`

```python
    def test_simple(self):
        def f(x):
            # simple test: 1 view op, 1 inplace op
            tmp = torch.ones(4, 2)
            y = x.view(4, 2)
            y.add_(tmp)
            z = x * x
            return y
        self.assert_functionalization(f, torch.ones(4, 2))
        logs = self.get_logs(f, torch.ones(4, 2))
        self.assertExpectedInline(logs, """\



def forward(self, arg0_1):
    ones = torch.ops.aten.ones.default([4, 2], device = device(type='cpu'), pin_memory = False)
    view_copy = torch.ops.aten.view_copy.default(arg0_1, [4, 2])
    add = torch.ops.aten.add.Tensor(view_copy, ones);  view_copy = ones = None
    view_copy_1 = torch.ops.aten.view_copy.default(add, [4, 2]);  add = None
    view_copy_2 = torch.ops.aten.view_copy.default(view_copy_1, [4, 2])
    mul = torch.ops.aten.mul.Tensor(view_copy_1, view_copy_1)
    copy_ = torch.ops.aten.copy_.default(arg0_1, view_copy_1);  arg0_1 = view_copy_1 = None
    return view_copy_2
    """)

changed to 

```python
def forward(self, arg0_1):
    ones = torch.ops.aten.ones.default([4, 2], device = device(type='cpu'), pin_memory = False)
    _param_constant0 = self._param_constant0
    detach = torch.ops.aten.detach.default(_param_constant0);  _param_constant0 = None
    view_copy = torch.ops.aten.view_copy.default(arg0_1, [4, 2])
    _param_constant1 = self._param_constant1
    detach_1 = torch.ops.aten.detach.default(_param_constant1);  _param_constant1 = None
    _param_constant2 = self._param_constant2
    detach_2 = torch.ops.aten.detach.default(_param_constant2);  _param_constant2 = None
    add = torch.ops.aten.add.Tensor(view_copy, ones);  view_copy = ones = None
    view_copy_1 = torch.ops.aten.view_copy.default(add, [4, 2]);  add = None
    view_copy_2 = torch.ops.aten.view_copy.default(view_copy_1, [4, 2])
    _param_constant3 = self._param_constant3
    detach_3 = torch.ops.aten.detach.default(_param_constant3);  _param_constant3 = None
    _param_constant4 = self._param_constant4
    detach_4 = torch.ops.aten.detach.default(_param_constant4);  _param_constant4 = None
    mul = torch.ops.aten.mul.Tensor(view_copy_1, view_copy_1)
    copy_ = torch.ops.aten.copy_.default(arg0_1, view_copy_1);  arg0_1 = view_copy_1 = None
    return view_copy_2
    """)
```

As can be seen, a bunch of 

```python
    _param_constantXXX = self._param_constantXXX
    detach = torch.ops.aten.detach.default(_param_constantXXX);  _param_constantXXX = None
```

show up. Any idea how to get around this?

ezyang replied:
Submit this by itself, without any other changes

ezyang commented:
sorry, i need to look at this carefully but haven't had time yet

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -444,7 +446,7 @@ def getattr(self, attr, attr_val, parameter_proxy_cache):
 
     def create_arg(self, a: Any):
-        if isinstance(a, torch.nn.Parameter):
+        if isinstance(a, (torch.nn.Parameter, torch._subclasses.FakeTensor)):
```
This I think is unnecessary, a properly fakeified parameter should still isinstance(x, Parameter)

thiagocrepaldi commented:
Related discussions:
https://github.com/pytorch/pytorch/issues/101288
https://github.com/pytorch/pytorch/issues/101289

Commit:
Add support to symbolic tracing to torch.export API

Summary of changes:

	* Plumb fake_mode through torch.export API. When specified, it
replaces the creation of a new FaketendorMode at InstructionTranslator on behalf of OutputGraph
	* Hacks FakeTensor.__new__ to prevent a
torch.tensor._make_subclass call for inputs that are already fakefied by
user. This probably need to be fixed in a nicer way. Any idea?
	* Removed a few asserts that didn't want faked tensors coming
from user script
	* Added torch._subclasses.fake_tensor.FakeTensor to type list on a few
asserts check to allow fake inputs

Commit:
Experimenting allowing pre-existing tensor for make_subclass

ezyang commented:
OK, so even after https://github.com/pytorch/pytorch/pull/102091 there is some trouble. The problem is in the make_fx call inside dynamo.export. The current behavior of dynamo.export with aten=True is to keep parameters directly on the exported module as attribute accesses. With the updated invariants this will fail, because you will have access to free fake parameters inside make_fx.

We can fix this by allowing make_fx to put free fake tensors as constants. But I am loathe to turn this on unconditionally, as it usually indicates bugs. In this case, maybe we can have it as an extra option which is true when fake_mode is non-None.

There's another possibility which is to use aot_module_export. But this is BC-breaking because all parameters become explicit inputs. But maybe you prefer this!

thiagocrepaldi commented:
> 

Thanks Edward, having parameters as inputs is tricky for final users and IMO we only should use it when we have no other choice.

If keeping free fake tensors as constants in the FX does not have any side-effect when fake mode is on, I think it should be ok to guard it behind a flag

thiagocrepaldi commented:
> > 
> 
> Thanks Edward, having parameters as inputs is tricky for final users and IMO we only should use it when we have no other choice. We want to be as close to as possible from the `torch.export` backend, so if it uses the `aot_module_export`, we might discuss it a bit further.
> 
> However, if keeping free fake tensors as constants in the FX does not have any side-effect when fake mode is on, I think it should be ok to guard it behind a flag

Hi Edward, is the error regarding free fake tensors manifesting in the unit test for this PR? See below (`TypeError: Multiple dispatch failed for 'torch._ops.aten.t.default'; all __torch_dispatch__ handlers returned NotImplemented:`)

```bash
test/export/test_export.py::TestExport::test_capture_symbolic_tracing FAILED [0.8830s]

================================================================================================================================== FAILURES ==================================================================================================================================
__________________________________________________________________________________________________________________ TestExport.test_capture_symbolic_tracing __________________________________________________________________________________________________________________
Traceback (most recent call last):
  File "/opt/pytorch/test/export/test_export.py", line 606, in test_capture_symbolic_tracing
    graph_module, _ = torchdynamo.export(
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1028, in export
    graph = make_fx(
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 788, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_autograd), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 287, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 484, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 287, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 810, in trace
    (self.create_arg(fn(*args)),),
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 501, in wrapped
    out = f(*tensors)
  File "<string>", line 1, in <lambda>
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 1024, in graph_with_interpreter
    return torch.fx.Interpreter(graph).run(*args)
  File "/opt/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/opt/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/opt/pytorch/torch/fx/interpreter.py", line 312, in call_module
    return submod(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 788, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 450, in call_module
    return forward(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 781, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1502, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/module.py", line 1511, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/pytorch/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
TypeError: Multiple dispatch failed for 'torch._ops.aten.t.default'; all __torch_dispatch__ handlers returned NotImplemented:

  - mode object <torch.fx.experimental.proxy_tensor.ProxyTorchDispatchMode object at 0x7fe8ac3c65e0>
  - tensor subclass <class 'torch._subclasses.fake_tensor.FakeTensor'>

For more information, try re-running with TORCH_LOGS=not_implemented

While executing %l__self___linear : [#users=1] = call_module[target=L__self___linear](args = (%l_x_,), kwargs = {})
Original traceback:
  File "/opt/pytorch/test/export/test_export.py", line 587, in forward
    out = self.linear(x)

----------------------------------------------------------------------------------------------------------------------------- Captured log call ------------------------------------------------------------------------------------------------------------------------------
ERROR    numba.cuda.cudadrv.driver:driver.py:405 Call to cuInit results in CUDA_ERROR_NO_DEVICE
========================================================================================================================== short test summary info ===========================================================================================================================
FAILED [0.8830s] test/export/test_export.py::TestExport::test_capture_symbolic_tracing - TypeError: Multiple dispatch failed for 'torch._ops.aten.t.default'; all __torch_dispatch__ handlers returned NotImplemented:
====================================================================================================================== 1 failed, 20 deselected in 7.33s ======================================================================================================================
(ptca) root@cd482011d639:/opt/pytorch# 
```

ezyang commented:
Yes, it's precisely that.

To allow for free fake tensors, we just need to add a new kwarg to `make_fx` which allows for FakeTensor in handled types (handled types correspond to things we are willing to inline as constants into the traced graph.

thiagocrepaldi commented:
> Yes, it's precisely that.
> 
> To allow for free fake tensors, we just need to add a new kwarg to `make_fx` which allows for FakeTensor in handled types (handled types correspond to things we are willing to inline as constants into the traced graph.

Any suggestion for the argument name? `make_fx` already has a `_allow_non_fake_inputs=False`. `_allow_free_fake_tensor=False`  might be an option, but it might be confusing given the `_allow_non_fake_inputs=False` already.

ezyang commented:
`_allow_fake_constant`

Commit:
Add allow_fake_constants argumento to make_fx

thiagocrepaldi commented:
> `_allow_fake_constant`

done, feel free to take another look. I have triggered `make_fx(allow_fake_constants=fake_mode is not None,...)` to prevent yet another flag to `torch._dynamo.export(fake_mode=user_fake_mode)`

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -615,7 +615,8 @@ static PyObject* THPVariable_make_subclass(
       std::move(data),
-      c10::impl::PyInterpreterStatus::DEFINITELY_UNINITIALIZED);
+      c10::impl::PyInterpreterStatus::DEFINITELY_UNINITIALIZED,
+      /*allow_preexisting_pyobj=*/true);
```
So you gonna revert this right

ezyang reviewed (commented):
Looks good, we should probably figure out if we actually wanna plumb fake mode or do something fancy like stuff it in torch._guards context so we can revert that part of the pr

BowenBao commented on test/export/test_export.py:
```diff
@@ -213,5 +213,47 @@ def f(x, y):
+
+        # Export the model with fake inputs and parameters
+        for aten_graph in [True, False]:
+            graph_module, _ = torchdynamo.export(
```
Is it required that this call happens outside of `fake_mode` context?

thiagocrepaldi replied:
Yes, perd Edward's request

BowenBao replied:
I see, is it because it's more in line with previous dynamo assumptions? I guess it'll be good to raise with explanation for dynamo if pre-existing fake mode context is detected (not asking to add in this PR), or maybe that's done already?

BowenBao commented on test/export/test_export.py:
```diff
@@ -213,5 +213,47 @@ def f(x, y):
 
 
+    @unittest.skipIf(not torchdynamo.is_dynamo_supported(), "dynamo doesn't support")
+    def test_capture_symbolic_tracing(self) -> None:
```
nit: feels `test/dynamo/test_export.py` is more suited location.

Commit:
Address comments

Commit:
lint

ezyang commented:
Yeah OK, reflecting on this some more, I want to do this through guards. I can make the adjustment to the PR if you want.

How are the CI failures looking?

thiagocrepaldi commented:
> Yeah OK, reflecting on this some more, I want to do this through TracingContext. I can make the adjustment to the PR if you want.
> 
> How are the CI failures looking?

Current solution partially uses `TracingContext` already. The `fake_mode` created by the user is passed to `torch._dynamo.export(fake_mode)`. 

The execution flow that takes `fake_mode` from user and initialize dynamo with it is: user instantiates `fake_mode` and pass it along to `torch._dynamo.export(fake_mode)` -> `optimize_assert(fake_mode)` -> `convert_frame.convert_frame_assert(fake_mode)` -> `_convert_frame_assert(fake_mode)` -> `_compile(fake_mode)` -> `transform_code_object(transform)` -> `transform(fake_mode)` -> `InstructionTranslator(fake_mode)` -> `OutputGraph(fake_mode)` -> `TracingContext(fake_mode)`.

Given that, what do you mean by "do this through TracingContext"? In a sense, it is already done by `TracingContext`. The thing is that `TracingContext` is only instantiate late in the initialization process, but `fake_mode` is instantiated before `torch._dynamo.export` is ever called.

Feel free to append a commit to the PR, but I can help with that too
```bash
+ docker build --no-cache --progress=plain --build-arg BUILD_ENVIRONMENT=308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7:2449172a5f2d2b623e0baab605db55c4105e35f5 --build-arg PROTOBUF=yes --build-arg THRIFT= --build-arg LLVMDEV= --build-arg DB=yes --build-arg VISION=yes --build-arg UBUNTU_VERSION=18.04 --build-arg CENTOS_VERSION= --build-arg DEVTOOLSET_VERSION= --build-arg GLIBC_VERSION= --build-arg CLANG_VERSION= --build-arg ANACONDA_PYTHON_VERSION=3 --build-arg GCC_VERSION=7 --build-arg CUDA_VERSION=11.8 --build-arg CUDNN_VERSION=8 --build-arg TENSORRT_VERSION= --build-arg ANDROID= --build-arg ANDROID_NDK= --build-arg GRADLE_VERSION= --build-arg VULKAN_SDK_VERSION= --build-arg SWIFTSHADER= --build-arg CMAKE_VERSION=3.18.5 --build-arg NINJA_VERSION= --build-arg KATEX= --build-arg ROCM_VERSION= --build-arg PYTORCH_ROCM_ARCH=gfx906 --build-arg IMAGE_NAME=nvidia/cuda:11.8-devel-ubuntu18.04 --build-arg UCX_COMMIT= --build-arg UCC_COMMIT= --build-arg CONDA_CMAKE= --build-arg TRITON= --build-arg ONNX= --build-arg DOCS= -f ubuntu-cuda/Dockerfile -t tmp.rjp07pwbip -t 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7:2449172a5f2d2b623e0baab605db55c4105e35f5:2449172a5f2d2b623e0baab605db55c4105e35f5 .
invalid argument "308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda11.8-cudnn8-py3-gcc7:2449172a5f2d2b623e0baab605db55c4105e35f5:2449172a5f2d2b623e0baab605db55c4105e35f5" for "-t, --tag" flag: invalid reference format
See 'docker build --help'.
Error: Process completed with exit code 125.
```

ezyang committed:
Get rid of plumbing for dynamic state

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
fixup

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang reviewed (approved):
OK pushed the fixups. This is GTG.

ezyang commented:
@pytorchbot merge

ezyang committed:
lint

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang commented:
Not sure what's up with the segfaults...

thiagocrepaldi reviewed (commented):
Thank you

thiagocrepaldi commented:
> Not sure what's up with the segfaults...

This one coming from this [assert](https://github.com/thiagocrepaldi/pytorch/blob/45bd0932afad09d049341cde1d59566991cabcd2/torch/_subclasses/fake_tensor.py#L1224)?

```bash
2023-06-05T13:25:24.7075383Z test_ops.py::TestFakeTensorCPU::test_pointwise_ops_signal_windows_blackman_cpu_float32 [2023-06-05 13:25:18,754] torch._dynamo.convert_frame: [ERROR] WON'T CONVERT __torch_dispatch__ /var/lib/jenkins/workspace/test/test_ops.py line 2097 
2023-06-05T13:25:24.7075452Z due to: 
2023-06-05T13:25:24.7075554Z Traceback (most recent call last):
2023-06-05T13:25:24.7075851Z   File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/_subclasses/fake_tensor.py", line 1224, in dispatch
2023-06-05T13:25:24.7075926Z     assert all(
2023-06-05T13:25:24.7076155Z torch._dynamo.exc.TorchRuntimeError: Failed running call_function aten.lift_fresh.default(*(FakeTensor(..., size=(3,)),), **{}):
2023-06-05T13:25:24.7076282Z f{func} should not have fake inputs without constants
2023-06-05T13:25:24.7076295Z 
2023-06-05T13:25:24.7076355Z from user code:
2023-06-05T13:25:24.7076520Z    File "/var/lib/jenkins/workspace/test/test_ops.py", line 2100, in __torch_dispatch__
2023-06-05T13:25:24.7076607Z     out = func(*args, **kwargs)
2023-06-05T13:25:24.7076613Z 
2023-06-05T13:25:24.7076785Z Set torch._dynamo.config.verbose=True or TORCHDYNAMO_VERBOSE=1 for more information
```

ezyang commented:
We're failing in dynamo config. One possibility is that our removal of the fake tensor asserts is causing tests to proceed into buggy dynamo code. We can test this by separating the assert removals into their own PR and testing on them.

ezyang commented:
No, I didn't manage to repro locally.

Commit:
Fix rebase

Commit:
Fix preserve_rng_state for FakeTensor

After discussing with Jason Ansel about this, he agred and combine both
`preserve_rng_state` into one that supports `FakeTensor`

thiagocrepaldi commented:
@ezyang CI is all green. Do you mind checking whether it is ok to make [this change](https://github.com/pytorch/pytorch/pull/100017/files#diff-b99ae8eb082046d6353fe4ee536a0d846e1007c7d3b21aefae468ded18fbf13eR597) on `preserve_rng_state`? That was the only change since your last approval

In short, I've wrapped it around `with torch.utils._python_dispatch._disable_current_modes():`, similar to what the (torch/_functorch/aot_autograd.py::preserve_rng_state)[https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L115-L127] has done

thiagocrepaldi commented on torch/_dynamo/utils.py:
```diff
@@ -595,15 +595,19 @@ def clone_inputs(example_inputs):
 
 @contextmanager
 def preserve_rng_state():
-    rng = torch.clone(torch.random.get_rng_state())
```
See https://github.com/pytorch/pytorch/pull/97738#issuecomment-1593436738 for a discussion

thiagocrepaldi replied:
After offline discussion with @jansel, he agreed we should combine both `preserve_rng_state` implementation into one that supports fake tensor

thiagocrepaldi commented:
@pytorchbot merge

thiagocrepaldi commented:
@pytorchbot merge -f "Gloo Distributed tests are randomly failing. Unrelated to this PR"

ConnollyLeon commented:
Can it be used with AOT_Autograd to capture a joint_forward_backward graph now?
